mlir/test/Quantum/DisentangleCNOTTest.mlir

```diff
@@ -0,0 +1,217 @@
+// Copyright 2024 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// RUN: quantum-opt %s --pass-pipeline="builtin.module(func.func(disentangle-CNOT))" --split-input-file --verify-diagnostics | FileCheck %s
+
+
+// Explicit unit tests for all CNOT disentangling table entries
+// Note that these tests reuse qubits for conciseness.
+
+// CHECK: func.func private @circuit()
+  func.func private @circuit() -> f64 {
+    %cst = arith.constant 1.230000e+00 : f64
+    %_ = quantum.alloc( 2) : !quantum.reg
+    %ZERO_0 = quantum.extract %_[ 0] : !quantum.reg -> !quantum.bit
+    %ZERO_1 = quantum.extract %_[ 1] : !quantum.reg -> !quantum.bit
+    %ONE_0 = quantum.custom "PauliX"() %ZERO_0 : !quantum.bit
+    %ONE_1 = quantum.custom "PauliX"() %ZERO_1 : !quantum.bit
+    %PLUS_0 = quantum.custom "Hadamard"() %ZERO_0 : !quantum.bit
+    %PLUS_1 = quantum.custom "Hadamard"() %ZERO_1 : !quantum.bit
+    %MINUS_0 = quantum.custom "Hadamard"() %ONE_0 : !quantum.bit
+    %MINUS_1 = quantum.custom "Hadamard"() %ONE_1 : !quantum.bit
+    %OTHERS_0 = quantum.custom "RX"(%cst) %ZERO_0 : !quantum.bit
+    %OTHERS_1 = quantum.custom "RX"(%cst) %ZERO_1 : !quantum.bit
+    // CHECK: [[ZERO_0:%.+]] = quantum.extract {{%.+}}[ 0] : !quantum.reg -> !quantum.bit
+    // CHECK: [[ZERO_1:%.+]] = quantum.extract {{%.+}}[ 1] : !quantum.reg -> !quantum.bit
+    // CHECK: [[ONE_0:%.+]] = quantum.custom "PauliX"() [[ZERO_0]] : !quantum.bit
+    // CHECK: [[ONE_1:%.+]] = quantum.custom "PauliX"() [[ZERO_1]] : !quantum.bit
+    // CHECK: [[PLUS_0:%.+]] = quantum.custom "Hadamard"() [[ZERO_0]] : !quantum.bit
+    // CHECK: [[PLUS_1:%.+]] = quantum.custom "Hadamard"() [[ZERO_1]] : !quantum.bit
+    // CHECK: [[MINUS_0:%.+]] = quantum.custom "Hadamard"() [[ONE_0]] : !quantum.bit
+    // CHECK: [[MINUS_1:%.+]] = quantum.custom "Hadamard"() [[ONE_1]] : !quantum.bit
+    // CHECK: [[OTHERS_0:%.+]] = quantum.custom "RX"({{%.+}}) [[ZERO_0]] : !quantum.bit
+    // CHECK: [[OTHERS_1:%.+]] = quantum.custom "RX"({{%.+}}) [[ZERO_1]] : !quantum.bit
```

Comment:
Are these checks meaningful to the test? (not saying they aren't, just that we should always try and focus checks on things affected by the pass)

Reply:
Yes, these checks are meaningful. They are producing the qubit values with known states in the lit test's variable pool, so the disentangled CNOTs can be checked as 
```
    %0:2 = quantum.custom "CNOT"() %ZERO_0, %OTHERS_1 : !quantum.bit, !quantum.bit
    %user_0:2 = quantum.custom "CRZ"(%cst) %0#0, %0#1 : !quantum.bit, !quantum.bit
    // CHECK-NOT: quantum.custom "CNOT"
    // CHECK: {{%.+}} = quantum.custom "CRZ"({{%.+}}) [[ZERO_0]], [[OTHERS_1]] : !quantum.bit, !quantum.bit
```

Consider this above case. The action of the pass _is_ that the input to the CNOT becomes the input to the user gate. Of course, we can have one variable-defining check line per CNOT test, but this would mean 25 such check lines instead of a public pool of 10 lines here.